pipeline_dp/pipeline_operations.py

```diff
@@ -86,6 +101,42 @@ def group_by_key(self, col, stage_name: str):
     def filter(self, col, fn, stage_name: str):
         return col | stage_name >> beam.Filter(fn)
 
+    def filter_by_key(self, col, public_partitions, data_extractors, stage_name: str):
+        class PartitionsFilterJoin(beam.DoFn):
+            def process(self, joined_data):
+                key, rest = joined_data
+                values, is_public = rest.get(VALUES), rest.get(IS_PUBLIC)
+
+                # TODO the Issue #4 says this is blocked on other tasks. Revisit
+                # this once unblocked
+                if not values:
+                    return
+
+                if is_public:
+                    for value in values:
+                        yield key, value
+
+        def has_public_partition_key(pk_val):
+            return pk_val[0] in public_partitions
+
+        # define constants for using as keys in CoGroupByKey
+        VALUES, IS_PUBLIC = 0, 1
+
+        if public_partitions is None:
+            raise TypeError("Must provide a valid public_partitions")
+
+        col = col | "Mapping data by partition" >> beam.Map(lambda x: (data_extractors.partition_extractor(x), x))
+
+        if isinstance(public_partitions, (list, set)):
+            if not isinstance(public_partitions, set):
+                public_partitions = set(public_partitions)
+            return col | "Filtering data from public partitions" >> beam.Filter(has_public_partition_key)
+        else:
+            public_partitions = public_partitions | "Creating public_partitions PCollection" >> beam.Map(lambda x: (x, True))
```

Comment:
Please add comment
\# Public partitions are not in memory. Filter out with a join.

Reply:
Done.